nydusd: add the config support of `amplify_io`

hangvane opened the pull request:
## Relevant Issue (if applicable)
_If there are Issues related to this PullRequest, please list it._

## Details
_Please describe the details of PullRequest._

The first commit is to rename the variables with similar usages into same names.

i.e.,

**`prefetch.batch_size`/`prefetch.merging_size` -> `batch_size` (filed name)/`prefetch_batch_size`(function name) (newly changed)**

`threads`/`threads_count` -> `threads_count`

`batch_size`/`amplify_io` ->`user_io_batch_size`

`bandwidth_rate`/`bandwidth_limit` -> `bandwidth_limit`

This change is only for the variables in nydusd runtime, and does not affect the filed names in the config file.

The second commit adds the support of `amplify_io` in the config file of nydusd to restrict read amplification.

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

imeoer commented on storage/src/factory.rs:
```diff
@@ -117,6 +117,7 @@ impl BlobFactory {
     ) -> IOResult<Arc<dyn BlobCache>> {
         let backend_cfg = config.get_backend_config()?;
         let cache_cfg = config.get_cache_config()?;
+        let ondemand_batch_size = config.get_rafs_config()?.batch_size;
```
Does this have a default value if we do not set it?

hangvane replied:
I'm not sure when the rafs config is None.
For runtime, it's not empty.
For build time, I didn't observe any failed integration test.
But the unit test `test_block_device()` will fail.
Seems the reason is the config is extracted from `BlobCacheMgr::get_meta_info()`,
https://github.com/dragonflyoss/nydus/blob/3bb124ba77d1deee830807bd48ccb7a9081a368b/service/src/blob_cache.rs#L319-L323
where `BlobCacheEntry` does not store rafs config,
https://github.com/dragonflyoss/nydus/blob/3bb124ba77d1deee830807bd48ccb7a9081a368b/api/src/config.rs#L953-L968
But I don't know when the config extracted from `BlobCacheEntry::get_meta_info()` is used to call `new_blob_cache()`.

Anyway, I changed this code to assume that rafs config can be None here, for passing the unit test `test_block_device()`.

hangvane replied:
`batch_size` does have a default value if it is not set in the config file.
https://github.com/dragonflyoss/nydus/blob/3bb124ba77d1deee830807bd48ccb7a9081a368b/api/src/config.rs#L818-L824

imeoer commented on api/src/config.rs:
```diff
@@ -1194,11 +1194,15 @@ fn default_work_dir() -> String {
 }
 
-fn default_prefetch_batch_size() -> usize {
+fn default_batch_size() -> usize {
```
Can we give a more precise name for these options, how about (not sure I'm understanding this correctly):

```
default_merging_size -> default_general_request_batch_size
default_prefetch_batch_size -> default_prefetch_request_batch_size
default_batch_size -> default_user_io_batch_size
```

hangvane replied:
`merging_size` is different with `batch_size`. It just used the default value of `batch_size`.
Another two are renamed now, together with their associated variable names and function names.

hangvane replied:
Seems `merging_size` is also a kind of `prefetch_batch_size`. I'll figure it out.

hangvane replied:
There's multiple variable names for `merging_size`:
https://github.com/dragonflyoss/nydus/blob/e069235b7a6c52ebeb1c9343275f2258ac1c8187/api/src/config.rs#L871-L885
Now I'm keeping `PrefetchConfigV2::batch_size`, `FsPrefetchControl::merging_size`, `BlobPrefetchConfig::merging_size` as they are for Backward compatibility (since they implemented `Deserialize` trait), keeping `AsyncPrefetchConfig::merging_size` for consistency. The last two, `default_prefetch_request_batch_size()`, `FileCacheEntry::prefetch_request_batch_size()`, are renamed.

ccx1024cc commented on api/src/config.rs:
```diff
@@ -820,7 +820,7 @@ pub struct RafsConfigV2 {
     pub mode: String,
     /// Batch size to read data from storage cache layer.
-    #[serde(default = "default_batch_size")]
+    #[serde(default = "default_user_io_batch_size")]
```
Compatibility is required.

hangvane replied:
This setting is used to set the default value function, and does not change the variable names in the config file. So there's no compatibility issue.

ccx1024cc commented on api/src/config.rs:
```diff
@@ -1194,11 +1200,11 @@ fn default_work_dir() -> String {
-pub fn default_batch_size() -> usize {
-    128 * 1024
+fn default_user_io_batch_size() -> usize {
+    1024 * 1024
```
Any analysis for this change?

hangvane replied:
No analysis, but just keep the value as what is hard coded in the project currently, 1M. Nothing changed actually. See line 293 `batch_size: RAFS_DEFAULT_CHUNK_SIZE`

https://github.com/dragonflyoss/nydus/blob/f55d2c948f85e22041bda18d0f3c8de11ca4b8b7/storage/src/cache/fscache/mod.rs#L265-L294

ccx1024cc commented on api/src/config.rs:
```diff
@@ -875,7 +875,13 @@ pub struct PrefetchConfigV2 {
+    /// [AsyncPrefetchConfig::merging_size],
+    /// [default_prefetch_request_batch_size()],
+    /// [FileCacheEntry::prefetch_request_batch_size()].
+    #[serde(default = "default_prefetch_request_batch_size")]
```
It means changes for user. Please update [doc](https://github.com/dragonflyoss/nydus/blob/master/docs/nydusd.md) + add [smoke test](https://github.com/dragonflyoss/nydus/blob/master/tests/smoke.rs) and unit tests

hangvane replied:
This setting is used to set the default value function name, and does not change the variable names in the config file. So there's no compatibility issue.

imeoer commented:
@jiangliu PTAL.

imeoer commented on api/src/config.rs:
```diff
@@ -820,8 +820,8 @@ pub struct RafsConfigV2 {
     /// Batch size to read data from storage cache layer.
-    #[serde(default = "default_batch_size")]
-    pub batch_size: usize,
+    #[serde(rename = "batch_size", default = "default_user_io_batch_size")]
```
Any docs need to be updated? mainly update the comment of the field.

hangvane replied:
Added to `docs/nydusd.md`
For anywhere else, it is not provided since it is an optional field.

imeoer commented on storage/src/factory.rs:
```diff
@@ -117,6 +117,10 @@ impl BlobFactory {
     ) -> IOResult<Arc<dyn BlobCache>> {
         let backend_cfg = config.get_backend_config()?;
         let cache_cfg = config.get_cache_config()?;
+        let user_io_batch_size = match config.get_rafs_config() {
```
Better to unwrap to default 0 here.

hangvane replied:
It is unwrapped to `default_user_io_batch_size()` now, if can't get `rafs_config`.
So now we set this value exactly the same as the master branch (set to hard coded 1M, for any case including the case that can't get `rafs_config`)

imeoer replied:
Maybe try `let user_io_batch_size = config.get_rafs_config().ok().unwrap_or_else(default_user_io_batch_size)`.

hangvane replied:
Changed to map_or_else
```rust
let user_io_batch_size = config
    .get_rafs_config()
    .map_or_else(|_| default_user_io_batch_size(), |v| v.user_io_batch_size)
    as u32;
```

imeoer commented on storage/src/cache/fscache/mod.rs:
```diff
@@ -290,7 +292,9 @@ impl FileCacheEntry {
-            batch_size: RAFS_DEFAULT_CHUNK_SIZE,
+            // If none, use default 0 since it's not used.
+            // e.g., at build time.
+            user_io_batch_size: mgr.user_io_batch_size.unwrap_or(0),
```
Can the user set it to 0? It seems we have a `default_user_io_batch_size`.

hangvane replied:
It is not wrapped by Option anymore.
Users can set it to zero if they set it in the config file.

imeoer commented on docs/nydusd.md:
```diff
@@ -130,6 +130,9 @@ We are working on enabling cloud-hypervisor support for nydus.
   "iostats_files": true,
   // Enable support of fs extended attributes
   "enable_xattr": false,
+  // Amplified request batch size to read data from storage cache layer for user IO
```
How about `Amplified user IO request batch size to read data from remote storage backend / local cache`

hangvane replied:
Changed.

imeoer commented on misc/configs/nydusd-config-v2.toml:
```diff
@@ -142,7 +142,7 @@ bandwidth_limit = 10000000
 # Filesystem metadata cache mode, "direct" or "cached". "direct" is almost what you want.
 mode = "direct"
-# Batch size to read data from storage cache layer, valid values: 0-0x10000000
+# Amplified request batch size to read data from storage cache layer for user IO, valid values: 0-0x10000000
```
Ditto.

hangvane replied:
Changed the comments in all 5 places.

imeoer reviewed (approved):
LGTM, Thanks!

jiangliu commented on api/src/config.rs:
```diff
@@ -681,10 +681,10 @@ impl CacheConfigV2 {
 
         if self.prefetch.enable {
-            if self.prefetch.batch_size > 0x10000000 {
+            if self.prefetch.prefetch_request_batch_size > 0x10000000 {
```
how about `prefetch_batch_size` instead?

hangvane replied:
There's at least three types of `batch_size` in nydus. Is it explicit enough to leave it as `batch_size`?

hangvane replied:
Reverted to `prefetch.batch_size`. Please reply again if there's some other problems.

jiangliu commented on storage/src/factory.rs:
```diff
@@ -117,6 +117,10 @@ impl BlobFactory {
         let cache_cfg = config.get_cache_config()?;
+        let user_io_batch_size = config
+            .get_rafs_config()
+            .map_or_else(|_| default_user_io_batch_size(), |v| v.user_io_batch_size)
```
Should be?
```
.map_or_else(|| default_user_io_batch_size(), |v| v.user_io_batch_size)
```

hangvane replied:
function is expected to take 1 argument, but it takes 0 arguments
expected function that takes 1 argument

jiangliu commented on storage/src/cache/cachedfile.rs:
```diff
@@ -300,19 +301,19 @@ impl FileCacheEntry {
 
-    fn prefetch_batch_size(&self) -> u64 {
-        if self.prefetch_config.merging_size < 0x2_0000 {
+    fn prefetch_request_batch_size(&self) -> u64 {
```
How about `prefetch_batch_size()`?

hangvane replied:
Changed.

hangvane committed:
nydusd: add the config support of `amplify_io`

Add the support of `amplify_io` in the config file of nydusd
to configure read amplification.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
rafs: rename variable names about prefetch configuration

Variable names about prefetch are confused currently.
So we merge variable names that have the same meaning,
while DO NOT affect the field names read from the configuration file.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>
